### PR TITLE
added hyperbolastic functions

### PR DESCRIPTION
**Pull request by mayankray2020**
* functions
  * Hyperbolastic functions are introduced.

### CODE REVIEW

File: sympy/functions/special/hyperbolastic_functions.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+from sympy.functions.elementary.exponential import exp
+from sympy.functions.elementary.miscellaneous import sqrt
+from sympy.core.sympify import sympify
+from sympy.core.function import Function, ArgumentIndexError
+
+class h1(Function):
+    r"""
+    The hyperbolastic rate equation of type I, denoted H1, is given by:
+
+        dP(x)/dx = P(x)/M*(M-P(x))(delta + theta/(1+x**2)**(1/2))
```
Comment (JSS95) [inline sympy/functions/special/hyperbolastic_functions.py:12]: Can you write the formula in LaTeX format? Please refer to the [documentation guide page](https://docs.sympy.org/latest/documentation-style-guide.html#sample-docstring).

File: sympy/functions/special/hyperbolastic_functions.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+from sympy.functions.elementary.miscellaneous import sqrt
+from sympy.core.sympify import sympify
+from sympy.core.function import Function, ArgumentIndexError
+
+class h1(Function):
+    r"""
+    The hyperbolastic rate equation of type I, denoted H1, is given by:
+
+    .. math::
+    \frac{dP(x)}{dx} = \frac{P(x)*(\M-P(x))*(\delta+\frac{\theta}{(1+x^2)^0.5})}{M}
```
Comment (JSS95) [inline sympy/functions/special/hyperbolastic_functions.py:13]: \frac{\mathrm{d}P(x)}{\mathrm{d}x}= \frac{P(x)}{M} \left(M-P \left(x\right)\right)\left(\delta+\frac{\theta}{\sqrt{1+x^2}}\right)

File: sympy/functions/special/hyperbolastic_functions.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+                alpha = (M - P0)/P0*asinh(exp(-delta*x0*gama))
+                return M/(1 + alpha*asinh(exp(-delta*x*gama))).simplify()
+
+
+class h3(Function):
+    r"""
+    The hyperbolastic rate equation of type III, denoted H3, is given by:
+
+        .. math::
+    \frac{dP(t)}{dt} = (M-P(t))*(\delta*\gamma*t^(\gamma-1)+\frac{theta}{(1+(\theta*t)^2)^0.5})
```
Comment (JSS95) [inline sympy/functions/special/hyperbolastic_functions.py:116]:
```suggestion
    .. math::
       \frac{\mathrm{d}P(x)}{\mathrm{d}x} = (M-P(t)) \left( \delta\gamma t^{(\gamma-1)}+\frac{\theta}{\sqrt{1+(\theta t)^2}} \right)
```

**Review by JSS95 — CHANGES_REQUESTED** [review]

File: sympy/functions/special/hyperbolastic_functions.py (hunk 2 above)
Comment (JSS95) [inline sympy/functions/special/hyperbolastic_functions.py:13]:
```suggestion
    .. math::
       \frac{\mathrm{d}P(x)}{\mathrm{d}x} = \frac{P(x)(M-P(x))\left(\delta+\frac{\theta}{\sqrt{1+x^2}}\right)}{M}
```

File: sympy/functions/special/hyperbolastic_functions.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    def eval(cls ,M, delta, theta, x0, P0 , x):
+        alpha = (M - P0)/P0*exp(delta*x0 - theta*asinh(x0))
+        return M/(1 + alpha*exp(-delta*x - theta*asinh(x))).simplify()
+
+class h2(Function):
+    r"""
+    The hyperbolastic rate equation of type II, denoted H2, is given by:
+
+    .. math::
+    \frac{dP(x)}{dx} = frac{\delta*\alpha*\gamma*P(x)^2*x^(\gamma-1)*tanh(\frac{M-P(x)}{\alpha*P(x)})}{M}
```
Comment (JSS95) [inline sympy/functions/special/hyperbolastic_functions.py:62]:
```suggestion
    .. math::
       \frac{\mathrm{d}P(x)}{\mathrm{d}x} = \frac{\delta\alpha\gamma P(x)^2 x^{\gamma-1}\tanh\left(\frac{M-P(x)}{\alpha P(x)}\right)}{M}
```

File: sympy/functions/special/hyperbolastic_functions.py (hunk 3 above)
Comment (JSS95) [inline sympy/functions/special/hyperbolastic_functions.py:116]:
```suggestion
    .. math::
       \frac{\mathrm{d}P(t)}{\mathrm{d}t} = (M-P(t))\left(\delta\gamma t^{\gamma-1}+\frac{\theta}{\sqrt{1+(\theta t)^2}}\right)
```

**Comment by JSS95** [comment]
I edited your release note, but we may need more detailed information. Also we need to add these classes to `doc/src/modules/functions/special.rst` file.

**Comment by mayankray2020** [comment]
> I edited your release note, but we may need more detailed information. Also we need to add these classes to `doc/src/modules/functions/special.rst` file.

@JSS95  https://en.wikipedia.org/wiki/Hyperbolastic_functions information about these functions are provided here and other references regarding these functions are also available there.

**Comment by mayankray2020** [comment]
@JSS95 added the chlasses to doc/src/modules/functions/special.rst file. Please review this.

**Review by JSS95 — APPROVED** [review]

File: sympy/functions/special/hyperbolastic_functions.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+class h1(Function):
+    r"""
+    The hyperbolastic rate equation of type I, denoted H1, is given by:
+
+    .. math::
+        \frac{\mathrm{d}P(x)}{\mathrm{d}x} = \frac{P(x)(M-P(x))\left(\delta+\frac{\theta}{\sqrt{1+x^2}}\right)}{M}
+
+    where x is any real number and P(x) is the population size at x.
+    The parameter M represents carrying capacity, and parameters delta
+    and theta  jointly represent growth rate. The parameter theta gives the
```
Comment (eric-wieser) [inline sympy/functions/special/hyperbolastic_functions.py:17]:
```suggestion
    and theta jointly represent growth rate. The parameter theta gives the
```

File: sympy/functions/special/hyperbolastic_functions.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+    distance from a symmetric sigmoidal curve.
+
+    Examples
+    ========
+
+    >>> from sympy import diff, symbols, h1
+    >>> a ,b, c, x, y, z = symbols('M delta theta x0 P0 x')
```
Comment (eric-wieser) [inline sympy/functions/special/hyperbolastic_functions.py:24]:
```suggestion
    >>> a, b, c, x, y, z = symbols('M delta theta x0 P0 x')
```

File: sympy/functions/special/hyperbolastic_functions.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+    curves with an asymptote at M and is negative only for decay profiles.
+    We refer to the growth rate model above as the hyperbolastic differential
+    equation of type II.
+
+
+    Examples
+    ========
+
+    >>> from sympy import h2, symbols
+    >>> a ,b, c, x, y, z = symbols('M delta gamma x0 P0 x')
```
Comment (eric-wieser) [inline sympy/functions/special/hyperbolastic_functions.py:76]:
```suggestion
    >>> a, b, c, x, y, z = symbols('M delta gamma x0 P0 x')
```

File: sympy/functions/special/hyperbolastic_functions.py [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+        else:
+            raise ArgumentIndexError(self, argindex)
+
+    @classmethod
+    def eval(cls ,M, delta, gama, x0, P0 , x):
+            if sympify(gama).is_real and gama <= 0 :
+                return ValueError("gamma must be greater than 0")
+            else:
+                alpha = (M - P0)/P0*asinh(exp(-delta*x0*gama))
+                return M/(1 + alpha*asinh(exp(-delta*x*gama))).simplify()
```
Comment (eric-wieser) [inline sympy/functions/special/hyperbolastic_functions.py:108]: Sloppy indents

File: sympy/functions/special/hyperbolastic_functions.py [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+            raise ArgumentIndexError(self, argindex)
+
+    @classmethod
+    def eval(cls ,M, delta, gama, theta, t0, P0 , t):
+            if sympify(t).is_real and t <= 0 :
+                return ValueError("t must be greater than 0")
+            else:
+
+                alpha = (M - P0)*exp(delta*t0**gama + asinh(theta*t0))
+                return M - alpha*exp(-delta*t**gama-asinh(theta*t)).simplify()
```
Comment (eric-wieser) [inline sympy/functions/special/hyperbolastic_functions.py:161]: As above

File: sympy/functions/special/hyperbolastic_functions.py [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+        if argindex == 6:
+            M, delta, theta, x0, P0 , x= self.args
+            p_x = h1(M, delta, theta, x0, P0 , x)
+            print(p_x)
+            return (p_x/M*(M - p_x)*(delta + theta/(sqrt(1 + x**2)))).simplify()
+        else:
+            raise ArgumentIndexError(self, argindex)
+
+    @classmethod
+    def eval(cls ,M, delta, theta, x0, P0 , x):
```
Comment (eric-wieser) [inline sympy/functions/special/hyperbolastic_functions.py:53]:
```suggestion
    def eval(cls, M, delta, theta, x0, P0 , x):
```

File: sympy/functions/special/hyperbolastic_functions.py [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+            M, delta, gama, x0, P0 , x= self.args
+            p_x = h2(M, delta, gama, x0, P0 , x)
+            alpha = (M - P0)/P0*asinh(exp(-delta*x0*gama))
+            return (p_x**2*x**(gama-1)*gama*delta*alpha)/M*tanh((M - p_x) \
+                                                    /alpha*p_x).simplify()
+        else:
+            raise ArgumentIndexError(self, argindex)
+
+    @classmethod
+    def eval(cls ,M, delta, gama, x0, P0 , x):
```
Comment (eric-wieser) [inline sympy/functions/special/hyperbolastic_functions.py:103]:
```suggestion
    def eval(cls, M, delta, gama, x0, P0 , x):
```

Comment (eric-wieser) [reply sympy/functions/special/hyperbolastic_functions.py:103]: Shouldn't this be `gamma` with two `m`s?

File: sympy/functions/special/hyperbolastic_functions.py [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+
+    def fdiff(self, argindex=1):
+        if argindex == 6:
+            M, delta, gama, x0, P0 , x= self.args
+            p_x = h2(M, delta, gama, x0, P0 , x)
+            alpha = (M - P0)/P0*asinh(exp(-delta*x0*gama))
+            return (p_x**2*x**(gama-1)*gama*delta*alpha)/M*tanh((M - p_x) \
+                                                    /alpha*p_x).simplify()
+        else:
+            raise ArgumentIndexError(self, argindex)
```
Comment (eric-wieser) [inline sympy/functions/special/hyperbolastic_functions.py:100]: Would `NotImplementedError` make more sense here, for when the index is valid but you don't know how to differentiate it?

**Review by eric-wieser — COMMENTED** [review]
Whitespace is fairly messy here - PEP8 says no spaces before commas.

File: sympy/functions/special/hyperbolastic_functions.py (hunk 12 above)
Comment (mayankray2020) [reply sympy/functions/special/hyperbolastic_functions.py]: @eric-wieser yes ig NotImplementedError would make more sense . I've made the suggested changes please review it.

**Comment by mayankray2020** [comment]
@oscarbenjamin @lglattly @eric-wieser please review this.

**Comment by oscarbenjamin** [comment]
I've never heard of hyperbolastic functions. It looks like they can be defined easily in terms of existing functions like `exp`. I'm not sure that we want to add these.

It looks like you can just do:
```python
h1 = 1/(1 + exp(-x-asinh(x)))
```

Can you explain what the advantage is having these functions rather than just using the existing functions like `exp`, `asinh` etc?

Sometimes there can be an advantage in adding functions even if they can be expressed using existing functions. Usually the reason would be that the functions have certain properties that can be exploited when manipulating them. I looked at the Wikipedia page and didn't see any particular properties:
https://en.wikipedia.org/wiki/Hyperbolastic_functions

**Comment by oscarbenjamin** [comment]
These functions certainly do not warrant being in the top-level sympy namespace with names like `h1`.

**Comment by mayankray2020** [comment]
> I've never heard of hyperbolastic functions. It looks like they can be defined easily in terms of existing functions like `exp`. I'm not sure that we want to add these.
> 
> It looks like you can just do:
> 
> ```python
> h1 = 1/(1 + exp(-x-asinh(x)))
> ```
> 
> Can you explain what the advantage is having these functions rather than just using the existing functions like `exp`, `asinh` etc?
> 
> Sometimes there can be an advantage in adding functions even if they can be expressed using existing functions. Usually the reason would be that the functions have certain properties that can be exploited when manipulating them. I looked at the Wikipedia page and didn't see any particular properties:
> https://en.wikipedia.org/wiki/Hyperbolastic_functions

These functions basically model problems like tumor growth, population growth, disease progression etc and can model both growth and decay curves. If we only use only exp or asinh which will act like a classical sigmoid curve and try to model biphasic growth models the accuracy would be very low. The main advantages of using hyperbolastic functions are that by varying the parameters the biphasic growth patterns can be modelled with much better accuracy than a classical sigmoid function can do. 



> These functions certainly do not warrant being in the top-level sympy namespace with names like `h1`.

That can be changed. Please suggest something which would sound better.

**Comment by eric-wieser** [comment]
@mayankray2020, I think you're missing @oscarbenjamin's point - they're asking why you would use `h1(x)` instead of `1/(1 + exp(-x-asinh(x)))` (which I assume is equal), not why you would use `h1(x)` instead of `exp(x)`.

**Comment by oscarbenjamin** [comment]
There was a similar discussion about adding sigmoid functions before https://github.com/sympy/sympy/pull/15952

**Comment by mayankray2020** [comment]
> @mayankray2020, I think you're missing @oscarbenjamin's point - they're asking why you would use `h1(x)` instead of `1/(1 + exp(-x-asinh(x)))` (which I assume is equal), not why you would use `h1(x)` instead of `exp(x)`.

`1/(1 + exp(-x-asinh(x)))`  is a classical sigmoid function which would not fit accurately on growth patterns with two phases of growth. Like in some pattern after a give time t the progression slows or reverse, in such pattern the sigmoid function would fail to fit the pattern after that time t. So in such cases of biphasic growth models (having two phases) h1 would prove to be fitting the given pattern much more accurately.

**Comment by oscarbenjamin** [comment]
It is already possible to use hyperbolastic functions in sympy:
```python
In [1]: M, alpha, delta, theta = symbols('M, alpha, delta, theta')

In [2]: h1 = M / (1 + alpha*exp(-delta*x-theta*asinh(x)))

In [3]: h1
Out[3]: 
           M            
────────────────────────
   -δ⋅x - θ⋅asinh(x)    
α⋅ℯ                  + 1
```
I don't see what benefit the code added in this PR gives.

**Comment by mayankray2020** [comment]
> There was a similar discussion about adding sigmoid functions before #15952



> It is already possible to use hyperbolastic functions in sympy:
> 
> ```python
> In [1]: M, alpha, delta, theta = symbols('M, alpha, delta, theta')
> 
> In [2]: h1 = M / (1 + alpha*exp(-delta*x-theta*asinh(x)))
> 
> In [3]: h1
> Out[3]: 
>            M            
> ────────────────────────
>    -δ⋅x - θ⋅asinh(x)    
> α⋅ℯ                  + 1
> ```
> 
> I don't see what benefit the code added in this PR gives.

yes ig this pr mainly for convenience. Ok i'll close this one.

**Pull request closed without merging**